src/main/java/com/yahoo/bullet/querying/partitioning/SimpleEqualityPartitioner.java

```diff
@@ -127,84 +133,66 @@ public SimpleEqualityPartitioner(BulletConfig config) {
         Map<String, String> values = getFieldValues(record);
         /*
          * Generate a truth table for all possible combinations of the fields when using the field value or not using
-         * an integer to represent a binary of fields.size() chars where each one represents to include or not the field
-         * For fields not present (NO_FIELD mapped), the final set de-dupes them with the binary combination that
-         * ignores them
+         * an integer to represent a binary of fields.size() chars where each one represents to include or not include
+         * the field. Note, fields that are not present are NULL-mapped. When not included, they are ANY-mapped.
          */
         return IntStream.range(0, 1 << fields.size()).mapToObj(i -> binaryToKey(i, values)).collect(Collectors.toSet());
     }
 
-    private static boolean hasNonANDLogicals(List<Clause> filters) {
-        return filters != null && filters.stream().anyMatch(SimpleEqualityPartitioner::hasNonANDLogicals);
-    }
-
-    private static boolean hasNonANDLogicals(Clause clause) {
-        if (clause instanceof FilterClause) {
-            return false;
-        }
-        return clause.getOperation() != Clause.Operation.AND || hasNonANDLogicals(((LogicalClause) clause).getClauses());
-    }
-
-    private void mapFieldToFilters(Clause clause, Map<String, List<FilterClause>> mapping) {
-        if (clause instanceof FilterClause) {
-            mapFieldToFilter((FilterClause) clause, mapping);
+    private void mapFieldsToValues(Expression expression, Map<String, Set<Object>> mapping) {
+        if (!(expression instanceof BinaryExpression)) {
             return;
         }
-        List<Clause> clauses = ((LogicalClause) clause).getClauses();
-        // Cannot have null non-AND logicals as it is checked for.
-        clauses.forEach(c -> this.mapFieldToFilters(c, mapping));
-    }
-
-    private void mapFieldToFilter(FilterClause clause, Map<String, List<FilterClause>> mapping) {
-        String field = clause.getField();
-        if (clause.getOperation() != Clause.Operation.EQUALS || !fieldSet.contains(field)) {
-            return;
-        }
-        List<FilterClause> list = mapping.getOrDefault(field, new ArrayList<>());
-        list.add(clause);
-        mapping.put(field, list);
-    }
-
-    private boolean hasInvalidFilterClauses(List<FilterClause> filters) {
-        if (filters == null || filters.size() != 1)  {
-            return true;
+        BinaryExpression binary = (BinaryExpression) expression;
+        if (binary.getOp() == Operation.AND) {
+            mapFieldsToValues(binary.getLeft(), mapping);
+            mapFieldsToValues(binary.getRight(), mapping);
+        } else if (binary.getOp() == Operation.EQUALS) {
+            if (binary.getLeft() instanceof FieldExpression && binary.getRight() instanceof ValueExpression) {
```

Comment:
Same code; can we just figure out which is the field and which is the value and not repeat this?